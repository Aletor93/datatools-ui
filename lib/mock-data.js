// @flow

import multi from '@conveyal/woonerf/store/multi'
import promise from '@conveyal/woonerf/store/promise'
import {middleware as fetchMiddleware} from '@conveyal/woonerf/fetch'
import clone from 'lodash/cloneDeep'
import {browserHistory} from 'react-router'
import {routerMiddleware, routerReducer as routing} from 'react-router-redux'
import {applyMiddleware, combineReducers, createStore} from 'redux'
import thunkMiddleware from 'redux-thunk'

import admin from './admin/reducers'
import {defaultState as defaultAdminOrganizationsState} from './admin/reducers/organizations'
import {defaultState as defaultAdminUsersState} from './admin/reducers/users'
import alerts from './alerts/reducers'
import {defaultState as defaultAlertsActiveState} from './alerts/reducers/active'
import {defaultState as defaultAlertsAlertsState} from './alerts/reducers/alerts'
import editor from './editor/reducers'
import {defaultState as defaultEditorDataState} from './editor/reducers/data'
import {defaultState as defaultEditorMapState} from './editor/reducers/mapState'
import {defaultState as defaultEditorSettingsState} from './editor/reducers/settings'
import {defaultState as defaultEditorTimetableState} from './editor/reducers/timetable'
import gtfs from './gtfs/reducers'
import {defaultState as defaultGtfsFilterState} from './gtfs/reducers/filter'
import {defaultState as defaultGtfsPatternState} from './gtfs/reducers/patterns'
import {defaultState as defaultGtfsRoutesState} from './gtfs/reducers/routes'
import {defaultState as defaultGtfsShapesState} from './gtfs/reducers/shapes'
import {defaultState as defaultGtfsStopsState} from './gtfs/reducers/stops'
import {defaultState as defaultGtfsTimetablesState} from './gtfs/reducers/timetables'
import {defaultState as defaultGtfsValidationState} from './gtfs/reducers/validation'
import * as gtfsPlusReducers from './gtfsplus/reducers'
import {defaultState as defaultGtfsPlusState} from './gtfsplus/reducers/gtfsplus'
import * as managerReducers from './manager/reducers'
import {defaultState as defaultManagerLanguagesState} from './manager/reducers/languages'
import {defaultState as defaultManagerProjectsState} from './manager/reducers/projects'
import {defaultState as defaultManagerStatusState} from './manager/reducers/status'
import {defaultState as defaultManagerUiState} from './manager/reducers/ui'
import {defaultState as defaultManagerUserState} from './manager/reducers/user'
import signs from './signs/reducers'
import {defaultState as defaultSignActiveState} from './signs/reducers/active'
import {defaultState as defaultSignSignsState} from './signs/reducers/signs'
import type {AppState} from './types/reducers'

const defaultManagerState = {
  languages: defaultManagerLanguagesState,
  projects: defaultManagerProjectsState,
  status: defaultManagerStatusState,
  ui: defaultManagerUiState,
  user: defaultManagerUserState
}

/**
 * Returns a cloned version of the initial load state of the redux store.
 */
export function getMockInitialState(): AppState {
  return clone({
    admin: {
      organizations: defaultAdminOrganizationsState,
      users: defaultAdminUsersState
    },
    alerts: {
      ...defaultAlertsActiveState,
      ...defaultAlertsAlertsState
    },
    ...defaultManagerState,
    editor: {
      data: defaultEditorDataState,
      // The editSettings state below represents the default state of redux-undo
      // which this reducer uses.
      editSettings: {
        future: [],
        present: defaultEditorSettingsState,
        index: 0,
        limit: 1,
        past: []
      },
      mapState: defaultEditorMapState,
      timetable: defaultEditorTimetableState
    },
    gtfs: {
      filter: defaultGtfsFilterState,
      patterns: defaultGtfsPatternState,
      routes: defaultGtfsRoutesState,
      shapes: defaultGtfsShapesState,
      stops: defaultGtfsStopsState,
      timetables: defaultGtfsTimetablesState,
      validation: defaultGtfsValidationState
    },
    gtfsplus: defaultGtfsPlusState,
    routing: {}, // react-router state
    signs: {
      ...defaultSignActiveState,
      ...defaultSignSignsState
    }
  })
}

/**
 * Returns a cloned version of the state or the redux store with a feed loaded
 * for editing in the editor.
 */
export function getMockEditorState (): AppState {
  const editorState = getMockInitialState()
  return clone(editorState)
}

/**
 * Creates a new store that has reducers that will update state.
 */
export function makeMockStore (initialState: AppState = getMockInitialState()) {
  return createStore(
    // $FlowFixMe not sure why this is giving a flow error
    combineReducers({
      routing,
      ...{
        ...managerReducers,
        admin,
        alerts,
        signs,
        ...gtfsPlusReducers,
        editor,
        gtfs
      }
    }),
    initialState,
    applyMiddleware(...[
      fetchMiddleware,
      multi,
      promise,
      thunkMiddleware
    ])
  )
}
