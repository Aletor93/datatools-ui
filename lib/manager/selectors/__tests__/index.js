// @flow

import clone from 'lodash/cloneDeep'

import {getActiveProject, getRouteData, getTimetableData} from '../'

import {mockInitialState} from '../../../mock-data'

describe('manager > selectors >', () => {
  describe('getActiveProject', () => {
    it('should get the project', () => {
      const mockProject = {
        id: '1'
      }
      const mockState = {
        projects: {
          active: mockProject.id,
          all: [mockProject]
        }
      }
      expect(getActiveProject(mockState)).toMatchSnapshot()
    })
  })

  // WARNING: these are unit tests where the output of the function being tested
  // is very dependent on downstream implementation
  describe('getTimetableData', () => {
    const mockTimetableData = {
      feed: {
        patterns: [
          {
            stops: [
              {stop_id: '15', stop_name: 'TECHNOLOGY & METRO'},
              {stop_id: '14', stop_name: '1ST & METRO @ LR STATION'}
            ],
            trips: [
              {
                direction_id: 1,
                pattern_id: '248',
                service_id: 'Weekdays',
                stop_times: [
                  {
                    arrival_time: 23580,
                    departure_time: 23580,
                    stop_id: '14',
                    stop_sequence: 1,
                    timepoint: null
                  },
                  {
                    arrival_time: null,
                    departure_time: null,
                    stop_id: '15',
                    stop_sequence: 2,
                    timepoint: null
                  }
                ],
                trip_headsign: '10 SANTA CLARA CALTRAIN VIA AIRPORT',
                trip_id: '2369269',
                trip_short_name: null
              },
              {
                direction_id: 1,
                pattern_id: '248',
                service_id: 'Weekdays',
                stop_times: [
                  {
                    arrival_time: 28980,
                    departure_time: 28980,
                    stop_id: '14',
                    stop_sequence: 1,
                    timepoint: null
                  },
                  {
                    arrival_time: null,
                    departure_time: null,
                    stop_id: '15',
                    stop_sequence: 2,
                    timepoint: null
                  }
                ],
                trip_headsign: '10 SANTA CLARA CALTRAIN VIA AIRPORT',
                trip_id: '2369292',
                trip_short_name: null
              }
            ]
          }
        ]
      }
    }

    it('should get the data when no data is fetched', () => {
      expect(getTimetableData(mockInitialState)).toMatchSnapshot()
    })

    it('should get the data when data is available', () => {
      const mockState = clone(mockInitialState)
      mockState.gtfs.timetables.data = mockTimetableData
      expect(getTimetableData(mockState)).toMatchSnapshot()
    })

    it('should get the data when data is available and timepoint filter is set', () => {
      const mockState = clone(mockInitialState)
      mockState.gtfs.filter.timepointFilter = true
      mockState.gtfs.timetables.data = mockTimetableData
      expect(getTimetableData(mockState)).toMatchSnapshot()
    })

    it('should get the data when data is available and show arrivals filter is set', () => {
      const mockState = clone(mockInitialState)
      mockState.gtfs.filter.showArrivals = true
      mockState.gtfs.timetables.data = mockTimetableData
      expect(getTimetableData(mockState)).toMatchSnapshot()
    })
  })

  // WARNING: these are unit tests where the output of the function being tested
  // is very dependent on downstream implementation
  describe('getRouteDetailsData', () => {
    it('should output blank array when no route data exists', () => {
      expect(getRouteData(mockInitialState)).toMatchSnapshot()
    })

    it('should output data when route data exists', () => {
      const mockState = clone(mockInitialState)
      mockState.gtfs.routes.routeDetails.data = {
        numRoutes: 1,
        routes: [
          {
            route_id: '10',
            route_short_name: '10',
            route_long_name: 'SANTA CLARA TRANSIT- METRO AIRPORT',
            route_desc: null,
            stops: [
              {
                stop_id: '20'
              },
              {
                stop_id: '5228'
              }
            ],
            trips: [
              {
                pattern_id: '284',
                stop_times: [
                  {
                    arrival_time: 23820,
                    departure_time: 23820
                  }
                ]
              }
            ]
          }
        ]
      }

      expect(getRouteData(mockState)).toMatchSnapshot()
    })
  })
})
