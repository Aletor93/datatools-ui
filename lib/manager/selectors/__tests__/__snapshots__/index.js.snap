// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager > selectors > getActiveProject should get the project 1`] = `
Object {
  "id": "1",
}
`;

exports[`manager > selectors > getRouteDetailsData should output blank array when no route data exists 1`] = `Array []`;

exports[`manager > selectors > getRouteDetailsData should output data when route data exists 1`] = `
Array [
  Object {
    "numPatterns": 1,
    "numStops": 2,
    "numTrips": 1,
    "routeId": "10",
    "routeName": "10 - SANTA CLARA TRANSIT- METRO AIRPORT",
    "tripsPerHour": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
]
`;

exports[`manager > selectors > getTimetableData should get the data when data is available 1`] = `
Object {
  "columns": Array [
    Object {
      "dataField": "tripId",
      "isKey": true,
      "name": "Trip ID",
      "row": "0",
      "rowSpan": "1",
      "width": 150,
    },
    Object {
      "dataField": "tripHeadsign",
      "name": "Trip Headsign",
      "row": "0",
      "rowSpan": "1",
      "width": 150,
    },
    Object {
      "dataField": "tripShortName",
      "name": "Trip Short Name",
      "row": "0",
      "rowSpan": "1",
      "width": 150,
    },
    Object {
      "colSpan": "1",
      "dataField": "stop-14-departure",
      "name": "1ST & METRO @ LR STATION",
      "row": "0",
      "width": 200,
    },
    Object {
      "colSpan": "1",
      "dataField": "stop-15-departure",
      "name": "TECHNOLOGY & METRO",
      "row": "0",
      "width": 200,
    },
  ],
  "rows": Array [
    Object {
      "stop-14-arrival": "06:33:00",
      "stop-14-arrival-raw": 23580,
      "stop-14-departure": "06:33:00",
      "stop-14-departure-raw": 23580,
      "stop-15-arrival": "",
      "stop-15-arrival-raw": null,
      "stop-15-departure": "",
      "stop-15-departure-raw": null,
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369269",
      "tripShortName": null,
    },
    Object {
      "stop-14-arrival": "08:03:00",
      "stop-14-arrival-raw": 28980,
      "stop-14-departure": "08:03:00",
      "stop-14-departure-raw": 28980,
      "stop-15-arrival": "",
      "stop-15-arrival-raw": null,
      "stop-15-departure": "",
      "stop-15-departure-raw": null,
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369292",
      "tripShortName": null,
    },
  ],
}
`;

exports[`manager > selectors > getTimetableData should get the data when data is available and show arrivals filter is set 1`] = `
Object {
  "columns": Array [
    Object {
      "dataField": "tripId",
      "isKey": true,
      "name": "Trip ID",
      "row": "0",
      "rowSpan": "2",
      "width": 150,
    },
    Object {
      "dataField": "tripHeadsign",
      "name": "Trip Headsign",
      "row": "0",
      "rowSpan": "2",
      "width": 150,
    },
    Object {
      "dataField": "tripShortName",
      "name": "Trip Short Name",
      "row": "0",
      "rowSpan": "2",
      "width": 150,
    },
    Object {
      "colSpan": "2",
      "dataField": undefined,
      "name": "1ST & METRO @ LR STATION",
      "row": "0",
      "width": 200,
    },
    Object {
      "dataField": "stop-14-arrival",
      "name": "Arrive",
      "row": "1",
      "width": 100,
    },
    Object {
      "dataField": "stop-14-departure",
      "name": "Depart",
      "row": "1",
      "width": 100,
    },
    Object {
      "colSpan": "2",
      "dataField": undefined,
      "name": "TECHNOLOGY & METRO",
      "row": "0",
      "width": 200,
    },
    Object {
      "dataField": "stop-15-arrival",
      "name": "Arrive",
      "row": "1",
      "width": 100,
    },
    Object {
      "dataField": "stop-15-departure",
      "name": "Depart",
      "row": "1",
      "width": 100,
    },
  ],
  "rows": Array [
    Object {
      "stop-14-arrival": "06:33:00",
      "stop-14-arrival-raw": 23580,
      "stop-14-departure": "06:33:00",
      "stop-14-departure-raw": 23580,
      "stop-15-arrival": "",
      "stop-15-arrival-raw": null,
      "stop-15-departure": "",
      "stop-15-departure-raw": null,
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369269",
      "tripShortName": null,
    },
    Object {
      "stop-14-arrival": "08:03:00",
      "stop-14-arrival-raw": 28980,
      "stop-14-departure": "08:03:00",
      "stop-14-departure-raw": 28980,
      "stop-15-arrival": "",
      "stop-15-arrival-raw": null,
      "stop-15-departure": "",
      "stop-15-departure-raw": null,
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369292",
      "tripShortName": null,
    },
  ],
}
`;

exports[`manager > selectors > getTimetableData should get the data when data is available and timepoint filter is set 1`] = `
Object {
  "columns": Array [
    Object {
      "dataField": "tripId",
      "isKey": true,
      "name": "Trip ID",
      "row": "0",
      "rowSpan": "1",
      "width": 150,
    },
    Object {
      "dataField": "tripHeadsign",
      "name": "Trip Headsign",
      "row": "0",
      "rowSpan": "1",
      "width": 150,
    },
    Object {
      "dataField": "tripShortName",
      "name": "Trip Short Name",
      "row": "0",
      "rowSpan": "1",
      "width": 150,
    },
    Object {
      "colSpan": "1",
      "dataField": "stop-14-departure",
      "name": "1ST & METRO @ LR STATION",
      "row": "0",
      "width": 200,
    },
  ],
  "rows": Array [
    Object {
      "stop-14-arrival": "06:33:00",
      "stop-14-arrival-raw": 23580,
      "stop-14-departure": "06:33:00",
      "stop-14-departure-raw": 23580,
      "stop-15-arrival": "",
      "stop-15-arrival-raw": null,
      "stop-15-departure": "",
      "stop-15-departure-raw": null,
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369269",
      "tripShortName": null,
    },
    Object {
      "stop-14-arrival": "08:03:00",
      "stop-14-arrival-raw": 28980,
      "stop-14-departure": "08:03:00",
      "stop-14-departure-raw": 28980,
      "stop-15-arrival": "",
      "stop-15-arrival-raw": null,
      "stop-15-departure": "",
      "stop-15-departure-raw": null,
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369292",
      "tripShortName": null,
    },
  ],
}
`;

exports[`manager > selectors > getTimetableData should get the data when no data is fetched 1`] = `
Object {
  "columns": Array [],
  "rows": Array [],
}
`;
